clamp/clamp.tex

```diff
@@ -0,0 +1,331 @@
+
+% Flipping $\din, \dout$: The worst case is equality. The relation accepts all greater distances. That's why the inequality is reversed. After chaining, the end-user can run a search (e.g., binary search) to tighten the $\dout$ down to the inequality.
+
+\smallskip
+
+%\silvia{If input metric equals output metric, do as above to avoid having to check that the end user passed them equal and raise exceptions?}
+
+%\silvia{Stability relation and defining $\din, \dout$.}
+
+%\silvia{Assuming it is ok to say bounded\_float?}
+
+% Mike: "You can use the total ordering construction-time check to justify that min and max in your function work with the argument and won't fail"
+
+% Also important: Programming framework setting $\mathcal{X} = \mathbb{R}$ is equivalent to $AllDomain<T>$ where $T$: Float.
+
+\section{Proof}
+The necessary definitions for the proof can be found at \href{https://www.overleaf.com/project/60d214e390b337703d200982}{``List of definitions used in the proofs"}.
+
+\subsection{Symmetric Distance}
+%\todo{Assumptions on min and max functions? Only min is defined in Rust, and assumes PartialOrd.}
+
+%\todo{The carry type in Rust; e.g., bounded R is f32. But since it is bounded, it must be between L and U, but this does not follow from f32.} 
+
+%\todo{Change theorem statement to include all needed  so that the theorems are claims about the constructors}
+
+%\silvia{Reference specific lines in the pseudocode? E.g., for function.}
+\begin{theorem}
+    For every setting of the input parameters \texttt{(L, U)} to \texttt{MakeClamp} such that the given preconditions
+    hold, the transformation returned by \texttt{MakeClamp} has the following properties:
+    \begin{enumerate}
+        \item \textup{(Appropriate output domain).} For every element $v$ in \texttt{input\_domain}, $\function(v)$ is in \texttt{output\_domain}. % Prof. Vadhan said on 29/6 to leave it as this for now, but maybe in the future we add to the theorem statement what exactly the input domain and output domain are (although of course we already say this in the proof)
+        
+        \item \textup{(Domain-metric compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ and $\dout$ are of type \texttt{u32}, if $v,w$ are $d_{in}$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $d_{out}$-close under \texttt{output\_metric}.
```

Comment:
A more general phrasing: "where din is the associated type for input_metric and dout is the associated type for output_metric".  That way you can use the same phrasing for all transformations.

Reply:
Yes this is better, we wanted a more general formualation -- changed.